shared/src/commonMain/kotlin/org/hyperskill/app/progresses/presentation/ProgressScreenReducer.kt

```diff
@@ -0,0 +1,133 @@
+                    InternalAction.LogAnalyticEvent(
+                        ProgressScreenViewedHyperskillAnalyticEvent()
+                    )
+                )
+            }
+        } ?: (state to emptySet())
+
+    private fun handleInitialize(state: State): ProgressScreenReducerResult? =
+        if (state.trackProgressState is TrackProgressState.Idle &&
+            state.projectProgressState is ProjectProgressState.Idle
+        ) {
+            state.copy(
+                trackProgressState = TrackProgressState.Loading,
+                projectProgressState = ProjectProgressState.Loading
+            ) to fetchContent()
+        } else {
+            null
+        }
+
+    private fun handlePullToRefresh(state: State): ProgressScreenReducerResult =
+        state.copy(
+            isTrackProgressRefreshing = true,
+            isProjectProgressRefreshing = true
+        ) to fetchContent(forceLoadFromNetwork = true) + setOf(
+            InternalAction.LogAnalyticEvent(
+                ProgressScreenClickedPullToRefreshHyperskillAnalyticEvent()
+            )
+        )
+
+    private fun handleRetryTrackProgressLoading(state: State): ProgressScreenReducerResult? =
+        if (state.trackProgressState is TrackProgressState.Error) {
+            state.copy(
+                trackProgressState = TrackProgressState.Loading
+            ) to setOf(
+                InternalAction.FetchTrackWithProgress(forceLoadFromNetwork = true)
+            )
+        } else {
+            null
+        }
+
+    private fun handleRetryProjectProgressLoading(state: State): ProgressScreenReducerResult? =
+        if (state.projectProgressState is ProjectProgressState.Error) {
+            state.copy(
+                projectProgressState = ProjectProgressState.Loading
+            ) to setOf(
+                InternalAction.FetchProjectWithProgress(forceLoadFromNetwork = true)
+            )
+        } else {
+            null
+        }
+
+    private fun handleTrackWithProgressFetchResult(
+        state: State,
+        message: ProgressScreenFeature.TrackWithProgressFetchResult
+    ): ProgressScreenReducerResult =
+        state.copy(
+            trackProgressState = when (message) {
+                ProgressScreenFeature.TrackWithProgressFetchResult.Error ->
+                    TrackProgressState.Error
+                is ProgressScreenFeature.TrackWithProgressFetchResult.Success ->
+                    TrackProgressState.Content(message.trackWithProgress)
+            },
+            isTrackProgressRefreshing = false
+        ) to emptySet()
+
+    private fun handleProjectWithProgressFetchResult(
+        state: State,
+        message: ProgressScreenFeature.ProjectWithProgressFetchResult
+    ): ProgressScreenReducerResult =
+        state.copy(
+            projectProgressState = when (message) {
+                ProgressScreenFeature.ProjectWithProgressFetchResult.Empty ->
+                    ProjectProgressState.Idle
```

Comment:
It's better to return `ProjectProgressState.Empty`, usually for idle states we show skeletons